File: web_backend/src/models/CompanyProduct.js
const mongoose = require('mongoose');

const companyProductSchema = new mongoose.Schema(
  {
    /** Set for one company only. Omit when [portfolioSuperAdminId] is set (all companies that super admin created). */
    companyId: { type: mongoose.Schema.Types.ObjectId, ref: 'Company', default: null, index: true },
    /** Partner/main super admin id — product visible to every company where Company.createdBySuperAdminId matches. */
    portfolioSuperAdminId: { type: mongoose.Schema.Types.ObjectId, ref: 'Admin', default: null, index: true },
    name: { type: String, required: true, trim: true },
    shortDescription: { type: String, trim: true, default: '' },
    fullDescription: { type: String, trim: true, default: '' },
    bannerImage: { type: String, trim: true, default: '' },
    /** Optional product video (https URL, e.g. mp4 or YouTube watch link). */
    videoUrl: { type: String, trim: true, default: '' },
    images: [{ type: String, trim: true }],
    price: { type: Number, default: null },
    offerTag: { type: String, trim: true, default: '' },
    showInApp: { type: Boolean, default: true },
    highlightProduct: { type: Boolean, default: false },
    showOnHomeBanner: { type: Boolean, default: false },
    status: { type: String, enum: ['active', 'inactive'], default: 'active' },
    ctaLabel: { type: String, trim: true, default: 'Contact Us' },
    ctaType: { type: String, enum: ['none', 'phone', 'url', 'email'], default: 'none' },
    ctaValue: { type: String, trim: true, default: '' },
  },
  { timestamps: true },
);

companyProductSchema.index({ companyId: 1, status: 1, showInApp: 1 });
companyProductSchema.index({ portfolioSuperAdminId: 1, status: 1, showInApp: 1 });

companyProductSchema.pre('validate', function companyProductScopeXor() {
  const hasCo = this.companyId != null;
  const hasP = this.portfolioSuperAdminId != null;
  if (!hasCo && !hasP) {
    this.invalidate('companyId', 'Set companyId or portfolioSuperAdminId.');
  }
  if (hasCo && hasP) {
    this.invalidate('companyId', 'Use only one of companyId or portfolioSuperAdminId.');
  }
});

module.exports = mongoose.model('CompanyProduct', companyProductSchema);

File: track/lib/models/company_product.dart
// track/lib/models/company_product.dart

class CompanyProductCard {
  const CompanyProductCard({
    required this.id,
    required this.name,
    required this.shortDescription,
    required this.bannerImage,
    required this.offerTag,
    required this.price,
    required this.highlightProduct,
    required this.showOnHomeBanner,
  });

  final String id;
  final String name;
  final String shortDescription;
  final String bannerImage;
  final String offerTag;
  final double? price;
  final bool highlightProduct;
  final bool showOnHomeBanner;

  factory CompanyProductCard.fromJson(Map<String, dynamic> j) {
    double? p;
    final raw = j['price'];
    if (raw is num) {
      p = raw.toDouble();
    } else if (raw is String && raw.trim().isNotEmpty) {
      p = double.tryParse(raw.trim());
    }
    return CompanyProductCard(
      id: '${j['id'] ?? j['_id'] ?? ''}',
      name: '${j['name'] ?? ''}',
      shortDescription: '${j['shortDescription'] ?? ''}',
      bannerImage: '${j['bannerImage'] ?? ''}',
      offerTag: '${j['offerTag'] ?? ''}',
      price: p,
      highlightProduct: j['highlightProduct'] == true,
      showOnHomeBanner: j['showOnHomeBanner'] == true,
    );
  }
}

class CompanyProductDetail extends CompanyProductCard {
  const CompanyProductDetail({
    required super.id,
    required super.name,
    required super.shortDescription,
    required super.bannerImage,
    required super.offerTag,
    required super.price,
    required super.highlightProduct,
    required super.showOnHomeBanner,
    required this.fullDescription,
    required this.images,
    required this.ctaLabel,
    required this.ctaType,
    required this.ctaValue,
  });

  final String fullDescription;
  final List<String> images;
  final String ctaLabel;
  final String ctaType;
  final String ctaValue;

  factory CompanyProductDetail.fromJson(Map<String, dynamic> j) {
    final base = CompanyProductCard.fromJson(j);
    final imgs = j['images'];
    final list = imgs is List
        ? imgs.map((e) => '$e').where((s) => s.isNotEmpty).toList()
        : <String>[];
    return CompanyProductDetail(
      id: base.id,
      name: base.name,
      shortDescription: base.shortDescription,
      bannerImage: base.bannerImage,
      offerTag: base.offerTag,
      price: base.price,
      highlightProduct: base.highlightProduct,
      showOnHomeBanner: base.showOnHomeBanner,
      fullDescription: '${j['fullDescription'] ?? ''}',
      images: list,
      ctaLabel: '${j['ctaLabel'] ?? 'Contact Us'}',
      ctaType: '${j['ctaType'] ?? 'none'}',
      ctaValue: '${j['ctaValue'] ?? ''}',
    );
  }
}

class CompanyProductHome {
  const CompanyProductHome({
    required this.banners,
    required this.highlighted,
  });

  final List<CompanyProductCard> banners;
  final List<CompanyProductCard> highlighted;

  static CompanyProductHome empty() =>
      const CompanyProductHome(banners: [], highlighted: []);

  factory CompanyProductHome.fromJson(Map<String, dynamic> j) {
    List<CompanyProductCard> parseList(dynamic raw) {
      if (raw is! List) return [];
      return raw
          .whereType<Map>()
          .map((e) => CompanyProductCard.fromJson(Map<String, dynamic>.from(e)))
          .toList();
    }

    return CompanyProductHome(
      banners: parseList(j['banners']),
      highlighted: parseList(j['highlighted']),
    );
  }
}
